[API design] Upload multiple versions of the same software to the same fleet (#45934)
fleetdm/fleet: docs/REST API/rest-api.md

      "conditional_access_enabled": false,
      "install_software": {
        "name": "Adobe Acrobat",
        "software_title_id": 1234,
        "installer_id": 23,

>         "package_name": "AdobeAcrobat-v24.1.pkg"

Reviewer: ```suggestion
        "installer_name": "AdobeAcrobat-v24.1.pkg"
```

    },
    "install_software": {
      "name": "Adobe Acrobat.app",
      "software_title_id": 1234,
      "installer_id": 23,

>       "package_name": "AdobeAcrobat-v24.1.pkg"

Reviewer: ```suggestion
      "installer_name": "AdobeAcrobat-v24.1.pkg"
```

    "labels_include_any": ["Macs on Sonoma"],
    "install_software": {
      "name": "Adobe Acrobat.app",
      "software_title_id": 1234,
      "installer_id": 23,

>       "package_name": "AdobeAcrobat-v24.1.pkg"

Reviewer: ```suggestion
      "installer_name": "AdobeAcrobat-v24.1.pkg"
```

---

> 

Reviewer: ```suggestion
```

- [Get software](#get-software)
- [Get software version](#get-software-version)
- [Get operating system version](#get-operating-system-version)
- [List packages for software title](#list-packages-for-software-title)
- [Add package](#add-package)

> - [Update package](#update-package)

Reviewer: ```suggestion
- [Add package](#add-package)
- [Update package](#update-package)
- [List packages for software](#list-packages-for-software)
```

| --------------- | ------- | ----- | ---------------------------------------------------------------------------------------------------------------------------------------------------- |
| id              | integer | path  | **Required.** The software title's ID.                                                                                                               |
| fleet_id        | integer | query | **Required**. The fleet ID. Lists packages for the specified fleet.                                                                                  |
| page            | integer | query | Page number of the results to fetch.                                                                                                                 |
| per_page        | integer | query | Results per page.                                                                                                                                    |
| order_key       | string  | query | What to order results by. Allowed fields are `name`, `version`, and `uploaded_at`. Default is `uploaded_at` (descending).                            |

> | order_direction | string  | query | **Requires `order_key`**. The direction of the order given the order key. Options include `"asc"` and `"desc"`. Default is `"desc"`.                 |

Reviewer: ```suggestion
```

Operating systems other than Windows, macOS, and Linux do not report vulnerabilities.

### List packages for software title

> 

Reviewer: ```suggestion
### List packages for software title

```

_Available in Fleet Premium._

> Returns the list of packages (installers) associated with a software title. A software title can have up to 5 packages. If multiple packages target the same host, Fleet will install the one added most recently.

Reviewer: ```suggestion
Returns the list of packages associated with a software title. A software title can have up to 5 packages for each fleet. If multiple packages target the same host, Fleet will install the one added most recently.
```

Operating systems other than Windows, macOS, and Linux do not report vulnerabilities.

### List packages for software title

> > **Experimental feature**. This feature is undergoing rapid improvement, which may result in breaking changes to the API or configuration surface. It is not recommended for use in automated workflows.

Reviewer: ```suggestion
```

        }
      ],
      "labels_include_all": null,
      "labels_exclude_any": null,
      "automatic_install_policies": [
        {
          "id": 343,
          "name": "[Install software] GlobalProtect",
          "type": "dynamic"
        }
      ],
      "status": {
        "installed": 323,
        "pending_install": 78,
        "failed_install": 31,
        "pending_uninstall": 0,
        "failed_uninstall": 0
      }
    },
    {
      "installer_id": 19,
      "name": "GlobalProtect-v6.3.2.pkg",
      "version": "6.3.2",
      "platform": "darwin",
      "uploaded_at": "2024-11-21T10:05:33Z",
      "hash_sha256": "abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789",
      "self_service": false,
      "fleet_maintained_app_id": null,
      "labels_include_any": null,
      "labels_include_all": null,
      "labels_exclude_any": null,
      "automatic_install_policies": null,
      "status": {
        "installed": 0,
        "pending_install": 0,
        "failed_install": 0,
        "pending_uninstall": 0,
        "failed_uninstall": 0
      }
    }
  ],
  "count": 2,
  "meta": {
    "has_next_results": false,
    "has_previous_results": false

>   }

Reviewer: ```suggestion
```

      "platform": "darwin",
      "uploaded_at": "2024-12-01T14:22:58Z",
      "hash_sha256": "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
      "self_service": true,
      "fleet_maintained_app_id": null,
      "labels_include_any": [
        {
          "name": "Engineering",
          "id": 294
        }
      ],
      "labels_include_all": null,
      "labels_exclude_any": null,
      "automatic_install_policies": [
        {
          "id": 343,
          "name": "[Install software] GlobalProtect",
          "type": "dynamic"
        }
      ],
      "status": {
        "installed": 323,
        "pending_install": 78,
        "failed_install": 31,
        "pending_uninstall": 0,
        "failed_uninstall": 0
      }
    },
    {
      "installer_id": 19,
      "name": "GlobalProtect-v6.3.2.pkg",
      "version": "6.3.2",
      "platform": "darwin",
      "uploaded_at": "2024-11-21T10:05:33Z",
      "hash_sha256": "abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789",
      "self_service": false,
      "fleet_maintained_app_id": null,
      "labels_include_any": null,
      "labels_include_all": null,
      "labels_exclude_any": null,

>       "target": "All hosts",

Reviewer: ```suggestion
```

Operating systems other than Windows, macOS, and Linux do not report vulnerabilities.

### List packages for software title

_Available in Fleet Premium._

> Returns the list of packages associated with a software title. A software title can have up to 5 packages for each fleet. If multiple packages target the same host, Fleet will install the one added most recently.

Reviewer: ```suggestion
Returns the list of packages associated with a software title. A software title can have up to 5 packages for each fleet. If multiple packages target the same host, Fleet will install the one added most recently (`uploaded_at`).
```

| id              | integer | path  | **Required.** The software title's ID.                                                                                                               |
| fleet_id        | integer | query | **Required**. The fleet ID. Lists packages for the specified fleet.                                                                                  |

#### Example

`GET /api/v1/fleet/software/titles/12/packages?fleet_id=3`

##### Default response

`Status: 200`

```json
{
  "packages": [
    {
      "installer_id": 23,
      "name": "GlobalProtect-v71.5.pkg",
      "version": "71.5",
      "platform": "darwin",
      "uploaded_at": "2024-12-01T14:22:58Z",
      "hash_sha256": "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
      "self_service": true,
      "fleet_maintained_app_id": null,
      "labels_include_any": [
        {
          "name": "Engineering",
          "id": 294
        }
      ],
      "labels_include_all": null,
      "labels_exclude_any": null,

>       "target": "Custom",

Reviewer: ```suggestion
```

| Name            | Type    | In    | Description                                                                                                                                          |
| --------------- | ------- | ----- | ---------------------------------------------------------------------------------------------------------------------------------------------------- |
| id              | integer | path  | **Required.** The software title's ID.                                                                                                               |
| fleet_id        | integer | query | **Required**. The fleet ID. Lists packages for the specified fleet.                                                                                  |

#### Example

`GET /api/v1/fleet/software/titles/12/packages?fleet_id=3`

##### Default response

`Status: 200`

```json
{
  "packages": [
    {
      "installer_id": 23,
      "name": "GlobalProtect-v71.5.pkg",
      "version": "71.5",
      "platform": "darwin",
      "uploaded_at": "2024-12-01T14:22:58Z",
      "hash_sha256": "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
      "self_service": true,

>       "fleet_maintained_app_id": null,

Reviewer: ```suggestion
```

      "labels_include_any": [
        {
          "name": "Engineering",
          "id": 294
        }
      ],
      "labels_include_all": null,
      "labels_exclude_any": null,
      "automatic_install_policies": [
        {
          "id": 343,
          "name": "[Install software] GlobalProtect",
          "type": "dynamic"
        }
      ],
      "status": {
        "installed": 323,
        "pending_install": 78,
        "failed_install": 31,
        "pending_uninstall": 0,
        "failed_uninstall": 0
      }
    },
    {
      "installer_id": 19,
      "name": "GlobalProtect-v6.3.2.pkg",
      "version": "6.3.2",
      "platform": "darwin",
      "uploaded_at": "2024-11-21T10:05:33Z",
      "hash_sha256": "abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789",
      "self_service": false,

>       "fleet_maintained_app_id": null,

Reviewer: ```suggestion
```
